src/Configuration/Config.AzureKeyVault/src/KeyVaultClientWrapper.cs

```diff
@@ -11,21 +15,64 @@ namespace Microsoft.Extensions.Configuration.AzureKeyVault
     /// <inheritdoc />
     internal class KeyVaultClientWrapper : IKeyVaultClient
     {
+        private sealed class EmptyPage : IPage<SecretItem>
+        {
+            public static readonly EmptyPage Instance = new EmptyPage();
+
+            public IEnumerator<SecretItem> GetEnumerator()
+            {
+                return Enumerable.Empty<SecretItem>().GetEnumerator();
+            }
+
+            IEnumerator IEnumerable.GetEnumerator()
+            {
+                return GetEnumerator();
+            }
+
+            public string NextPageLink { get; } = null;
+        }
+
         private readonly KeyVaultClient _keyVaultClientImplementation;
+        private readonly bool _optional;
 
         /// <summary>
         /// Creates a new instance of <see cref="KeyVaultClientWrapper"/>.
         /// </summary>
         /// <param name="keyVaultClientImplementation">The <see cref="KeyVaultClient"/> instance to wrap.</param>
-        public KeyVaultClientWrapper(KeyVaultClient keyVaultClientImplementation)
+        /// <param name="optional">Whether GetSecretsAsync returns an empty page instead of throwing when retrieval fails.</param>
+        public KeyVaultClientWrapper(KeyVaultClient keyVaultClientImplementation, bool optional)
         {
             _keyVaultClientImplementation = keyVaultClientImplementation;
+            _optional = optional;
         }
 
         /// <inheritdoc />
-        public Task<IPage<SecretItem>> GetSecretsAsync(string vault)
+        public async Task<IPage<SecretItem>> GetSecretsAsync(string vault)
         {
-            return _keyVaultClientImplementation.GetSecretsAsync(vault);
+            try
+            {
+                return await _keyVaultClientImplementation.GetSecretsAsync(vault);
+            }
+            catch (System.Net.Http.HttpRequestException)
+            {
+                //Host unknown or not reachable.
+                if (!_optional)
+                    throw;
+            }
+            catch (Azure.Services.AppAuthentication.AzureServiceTokenProviderException)
```

Comment:
These are invalid config conditions, not a problem with availability.

Reply:
Right, not a problem with availability, but not necessarily invalid config conditions either I think. I get this exception when running code locally and I haven't logged into Azure with "az login". So it's not an invalid config condition but...maybe the scenario here is "I want my code to run even though I haven't logged into Azure".